File: src/doctor.ts
import { access } from "node:fs/promises";
import { constants } from "node:fs";
import { AppConfig } from "./config.js";
import { commandExists, remotePathExpr, runSsh, shellQuote } from "./ssh.js";

interface Check {
  label: string;
  ok: boolean;
  detail: string;
  required: boolean;
}

export async function runDoctor(config: AppConfig, agentCommand: string): Promise<boolean> {
  const checks: Check[] = [];

  checks.push({
    label: "local platform",
    ok: process.platform === "darwin",
    detail: process.platform === "darwin" ? "macOS" : `${process.platform}; macOS is required for clipboard capture`,
    required: true,
  });

  for (const command of ["ssh", "scp", "swift"]) {
    const path = await commandExists(command);
    checks.push({
      label: `local ${command}`,
      ok: path !== null,
      detail: path ?? "not found in PATH",
      required: true,
    });
  }

  checks.push(await remoteCheck(config, "ssh access", "printf ok", "ok", true));
  checks.push(await remoteCommandCheck(config, "remote tmux", "tmux", true));
  checks.push(await remoteCommandCheck(config, "remote agent", agentCommand, true));
  checks.push(await remoteWritableCacheCheck(config));

  for (const check of checks) {
    const mark = check.ok ? "ok" : check.required ? "fail" : "warn";
    console.log(`${mark.padEnd(4)} ${check.label}: ${check.detail}`);
  }

  return checks.every((check) => check.ok || !check.required);
}

async function remoteCommandCheck(config: AppConfig, label: string, command: string, required: boolean): Promise<Check> {
  const result = await runSsh(config.host, `command -v ${shellQuote(command)} 2>/dev/null || true`);
  const detail = result.stdout.trim();
  return {
    label,
    ok: result.exitCode === 0 && detail.length > 0,
    detail: detail.length > 0 ? detail : `${command} not found on ${config.host}`,
    required,
  };
}

async function remoteWritableCacheCheck(config: AppConfig): Promise<Check> {
  const command = [
    `mkdir -p ${remotePathExpr(config.remoteCacheDir)}`,
    `test -d ${remotePathExpr(config.remoteCacheDir)}`,
    `test -w ${remotePathExpr(config.remoteCacheDir)}`,
    `printf ok`,
  ].join(" && ");
  return remoteCheck(config, "remote cache dir", command, "ok", true, config.remoteCacheDir);
}

async function remoteCheck(
  config: AppConfig,
  label: string,
  command: string,
  expected: string,
  required: boolean,
  successDetail = expected,
): Promise<Check> {
  const result = await runSsh(config.host, command);
  const out = result.stdout.trim();
  const err = result.stderr.trim();
  const ok = result.exitCode === 0 && out === expected;
  return {
    label,
    ok,
    detail: ok ? successDetail : err || out || `remote command exited ${result.exitCode}`,
    required,
  };
}

export async function assertReadable(path: string): Promise<boolean> {
  try {
    await access(path, constants.R_OK);
    return true;
  } catch {
    return false;
  }
}

File: src/doctor.rs
use anyhow::Result;

use crate::config::AppConfig;
use crate::ssh::{command_exists, remote_path_expr, run_ssh, shell_quote};

struct Check {
    label: String,
    ok: bool,
    detail: String,
    required: bool,
}

pub fn run_doctor(config: &AppConfig, agent_command: &str) -> Result<bool> {
    let mut checks = Vec::new();
    checks.push(Check {
        label: "local platform".to_string(),
        ok: cfg!(target_os = "macos"),
        detail: if cfg!(target_os = "macos") {
            "macOS".to_string()
        } else {
            "macOS is required for clipboard capture".to_string()
        },
        required: true,
    });

    for command in ["ssh", "swift"] {
        let path = command_exists(command)?;
        checks.push(Check {
            label: format!("local {command}"),
            ok: path.is_some(),
            detail: path.unwrap_or_else(|| "not found in PATH".to_string()),
            required: true,
        });
    }

    checks.push(remote_check(
        config,
        "ssh access",
        "printf ok",
        "ok",
        true,
        "ok",
    )?);
    checks.push(remote_command_check(config, "remote tmux", "tmux", true)?);
    checks.push(remote_command_check(
        config,
        "remote agent",
        agent_command,
        true,
    )?);
    checks.push(remote_writable_cache_check(config)?);

    for check in &checks {
        let mark = if check.ok {
            "ok"
        } else if check.required {
            "fail"
        } else {
            "warn"
        };
        println!("{mark:<4} {}: {}", check.label, check.detail);
    }

    Ok(checks.iter().all(|check| check.ok || !check.required))
}

fn remote_command_check(
    config: &AppConfig,
    label: &str,
    command: &str,
    required: bool,
) -> Result<Check> {
    let result = run_ssh(
        config,
        &format!("command -v {} 2>/dev/null || true", shell_quote(command)),
        None,
    )?;
    let detail = result.stdout.trim().to_string();
    Ok(Check {
        label: label.to_string(),
        ok: result.exit_code == 0 && !detail.is_empty(),
        detail: if detail.is_empty() {
            format!("{command} not found")
        } else {
            detail
        },
        required,
    })
}

fn remote_writable_cache_check(config: &AppConfig) -> Result<Check> {
    let dir = remote_path_expr(&config.remote_cache_dir);
    let command = format!("mkdir -p {dir} && test -d {dir} && test -w {dir} && printf ok");
    remote_check(
        config,
        "remote cache dir",
        &command,
        "ok",
        true,
        &config.remote_cache_dir,
    )
}

fn remote_check(
    config: &AppConfig,
    label: &str,
    command: &str,
    expected: &str,
    required: bool,
    success_detail: &str,
) -> Result<Check> {
    let result = run_ssh(config, command, None)?;
    let stdout = result.stdout.trim();
    let stderr = result.stderr.trim();
    let ok = result.exit_code == 0 && stdout == expected;
    Ok(Check {
        label: label.to_string(),
        ok,
        detail: if ok {
            success_detail.to_string()
        } else if !stderr.is_empty() {
            stderr.to_string()
        } else if !stdout.is_empty() {
            stdout.to_string()
        } else {
            format!("remote command exited {}", result.exit_code)
        },
        required,
    })
}
